src/merkle_tree/mod.rs

```diff
@@ -1,253 +1,399 @@
+        path.reverse();
+
+        Ok(Path {
+            leaf_index: index,
+            auth_path: path,
+            leaf_sibling_hash,
+        })
+    }
+
+    /// Given the index and new leaf, return an updated path in order from root to bottom non-leaf level
+    fn updated_path<L: ToBytes>(
+        &self,
+        index: usize,
+        new_leaf: &L,
+    ) -> Result<(LeafDigest<P>, Vec<TwoToOneDigest<P>>), crate::Error> {
+        // calculate the hash of leaf
+        let new_leaf_hash =
+            P::LeafHash::evaluate(&self.leaf_hash_param, &ark_ff::to_bytes!(&new_leaf)?)?;
+
+        // calculate leaf sibling hash and locate its position (left or right)
+        let (leaf_left, leaf_right) = if index & 1 == 0 {
+            // leaf on left
+            (&new_leaf_hash, &self.leaf_nodes[index + 1])
         } else {
-            Ok(Path { path })
+            (&self.leaf_nodes[index - 1], &new_leaf_hash)
+        };
+
+        // calculate the updated hash at bottom non-leaf-level
+        let mut path_bottom_to_top = Vec::with_capacity(self.height - 1);
+        {
+            path_bottom_to_top.push(P::TwoToOneHash::evaluate_both(
+                &self.two_to_one_hash_param,
+                &ark_ff::to_bytes!(&leaf_left)?,
+                &ark_ff::to_bytes!(&leaf_right)?,
+            )?);
+        }
+
+        // then calculate the updated hash from bottom to root
+        let leaf_index_in_tree = convert_index_to_last_level(index, self.height);
+        let mut prev_index = parent(leaf_index_in_tree).unwrap();
+        while !is_root(prev_index) {
+            let (left_hash_bytes, right_hash_bytes) = if is_left_child(prev_index) {
+                (
+                    ark_ff::to_bytes!(path_bottom_to_top.last().unwrap())?,
+                    ark_ff::to_bytes!(&self.non_leaf_nodes[sibling(prev_index).unwrap()])?,
+                )
+            } else {
+                (
+                    ark_ff::to_bytes!(&self.non_leaf_nodes[sibling(prev_index).unwrap()])?,
+                    ark_ff::to_bytes!(path_bottom_to_top.last().unwrap())?,
+                )
+            };
+            path_bottom_to_top.push(P::TwoToOneHash::evaluate_both(
+                &self.two_to_one_hash_param,
+                &left_hash_bytes,
+                &right_hash_bytes,
+            )?);
+            prev_index = parent(prev_index).unwrap();
         }
+
+        debug_assert_eq!(path_bottom_to_top.len(), self.height - 1);
+        let path_top_to_bottom: Vec<_> = path_bottom_to_top.into_iter().rev().collect();
+        return Ok((new_leaf_hash, path_top_to_bottom));
     }
-}
 
-#[derive(Debug)]
-pub enum Error {
-    IncorrectLeafIndex(usize),
-    IncorrectPathLength(usize),
-}
+    /// Update the leaf at `index` to updated leaf.
+    /// ```tree_diagram
+    ///         [A]
+    ///        /   \
+    ///      [B]    C
+    ///     / \   /  \
+    ///    D [E] F    H
+    ///   .. / \ ....
+    ///    [I] J
+    /// ```
```

Comment:
Nice tree diagram!

```suggestion
    /// Update the leaf at `index` to updated leaf. e.g. for the following tree of 8 leaves:
    /// ```tree_diagram
    ///         [A]
    ///        /   \
    ///      [B]    C
    ///     / \   /  \
    ///    D [E] F    H
    ///   .. / \ ....
    ///    [I] J
    /// ```
    /// update(3, {new leaf}) would swap the leaf value associated at `[I]` and cause a recomputation of `[A], [B], [E]`.
```

Reply:
Thanks for the added comments below the tree diagram! Side story: I temporarily put a tree diagram there for my own interest (it helps me to write the code of `update`) and accidentally forgot to delete it. This diagram seems to be good, so I will keep it anyway (along with the great comments below)